### DISCUSSION TITLE
[Scalability] 500+ concurrent SSH proxies?

### DISCUSSION THREAD

**Discussion by newts** [abhinavsingh/proxy.py#634] [Q&A] [answered]
This package looks great.

I am looking for a way to proxy different incoming connections to many (let's say a top limit of 500) tunneled remote hosts on different ports.  Without spawning 500 separate instances can proxy.py be used to handle many connections?

Probably map ports 9xxx to 7xxx tunneled ports or something similar.

Thanks!

**Comment by abhinavsingh** [comment, marked as answer]
Last benchmark for `proxy.py` says it can pull off 16k connections per second.

<img width="267" alt="Screen Shot 2021-10-26 at 10 59 53 PM" src="https://user-images.githubusercontent.com/126065/138930654-cabd9939-0867-4bff-a2fe-906c60d61a16.png">

With real world data served and upstream proxies, throughput will depend upon latency of upstream connection stream, but we should still be able to achieve multi-thousands of connection per second.

So for 500 clients, until they all start making 20-30 request per second concurrently, a single `proxy.py` instance can serve them well.

Talking about `SSH` proxies, there is an inbuilt support for that.  I remember it used to work e2e when I tested last, but wasn't well polished/integrated/documented due to `mypy` typings.   If we ignore typing warning, SSH proxies worked fine IIRC. 

`proxy.py` then acts like a local `ngrok` server with `SSH` proxies enabled.

Let me know how it goes.  Happy to jump back into it if necessary.

**Reply by newts** [reply]
Thanks!  that's great news on performance.  to be clear though I need one public facing server that will proxy incoming connections to many devices tunneled in to the server.   500 clients (possible) being proxied to 500 tunneled devices.  Not sure how they would sort out except each device has it's own tunnel port associated to it's own incoming port so proxy.py would need to manage 500 of these associations.  I didn't see how to configure this setup.
Also I need another 500 tunnels/incoming ports for websockets proxy.
Yes, it does look a lot like a ngrok where it looks like they use subdomains in the url to specify the individual proxy.  That could also be a solution if they can also do the websocket.

jv

**Reply by abhinavsingh** [reply]
@newts Apologies for no show for few weeks.  Getting back to it, I think what you described can be architected using `proxy.py`.  Note that `proxy.py` internally also supports to listen on multiple ports.  Or you could simply start multiple containers on different port.  Or you could use `ProxyPoolPlugin` and implement your own routing logic, instead of maintaining 1-to-1 instances for each proxy user.

**Reply by abhinavsingh** [reply]
On a newer Mac performance went up-to 39k requests per second :)

<img width="562" alt="Screen Shot 2022-01-23 at 1 35 50 AM" src="https://user-images.githubusercontent.com/126065/150653841-c05e4611-2979-4dd9-8b94-c69eddd7b348.png">
